File: backend/models/booking.py
"""
Booking Model
SQLAlchemy model for booking and reservation management
"""

from sqlalchemy import Column, String, Integer, Boolean, DateTime, Text, ForeignKey, Float
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid

from database import Base

class Booking(Base):
    """Booking model for reservations and check-ins"""
    
    __tablename__ = "bookings"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Property and room relationships
    property_id = Column(String, ForeignKey("hospitality_properties.id"), nullable=False)
    room_id = Column(String, ForeignKey("rooms.id"), nullable=False)
    
    # Guest information
    guest_name = Column(String, nullable=False)
    guest_email = Column(String, nullable=False)
    guest_phone = Column(String, nullable=True)
    
    # Booking dates
    check_in = Column(DateTime, nullable=False)
    check_out = Column(DateTime, nullable=False)
    
    # Guest count
    adults = Column(Integer, default=1, nullable=False)
    children = Column(Integer, default=0, nullable=False)
    
    # Pricing
    room_rate = Column(Float, nullable=False)
    total_amount = Column(Float, nullable=False)
    
    # Status
    status = Column(String, default="confirmed", nullable=False)  # pending, confirmed, checked_in, checked_out, cancelled, no_show
    payment_status = Column(String, default="pending", nullable=False)  # pending, paid, partial, refunded, failed
    
    # Additional information
    special_requests = Column(Text, nullable=True)
    cancellation_reason = Column(Text, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    checked_in_at = Column(DateTime, nullable=True)
    checked_out_at = Column(DateTime, nullable=True)
    cancelled_at = Column(DateTime, nullable=True)
    
    # Relationships
    property = relationship("HospitalityProperty")
    room = relationship("Room")
    
    def __repr__(self):
        return f"<Booking(id='{self.id}', guest='{self.guest_name}', check_in='{self.check_in}')>"
    
    def to_dict(self):
        """Convert booking to dictionary"""
        return {
            "id": self.id,
            "property_id": self.property_id,
            "room_id": self.room_id,
            "guest_name": self.guest_name,
            "guest_email": self.guest_email,
            "guest_phone": self.guest_phone,
            "check_in": self.check_in.isoformat() if self.check_in else None,
            "check_out": self.check_out.isoformat() if self.check_out else None,
            "adults": self.adults,
            "children": self.children,
            "room_rate": self.room_rate,
            "total_amount": self.total_amount,
            "status": self.status,
            "payment_status": self.payment_status,
            "special_requests": self.special_requests,
            "cancellation_reason": self.cancellation_reason,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "checked_in_at": self.checked_in_at.isoformat() if self.checked_in_at else None,
            "checked_out_at": self.checked_out_at.isoformat() if self.checked_out_at else None,
            "cancelled_at": self.cancelled_at.isoformat() if self.cancelled_at else None
        }
    
    def get_length_of_stay(self) -> int:
        """Calculate length of stay in nights"""
        if not self.check_in or not self.check_out:
            return 0
        
        stay_duration = self.check_out - self.check_in
        return max(0, stay_duration.days)
    
    def is_active(self) -> bool:
        """Check if booking is active (confirmed or checked in)"""
        return self.status in ['confirmed', 'checked_in']
    
    def can_be_cancelled(self) -> bool:
        """Check if booking can be cancelled"""
        return self.status in ['pending', 'confirmed']
    
    def can_be_modified(self) -> bool:
        """Check if booking can be modified"""
        return self.status in ['pending', 'confirmed']

class BookingPayment(Base):
    """Booking payment model"""
    
    __tablename__ = "booking_payments"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Booking relationship
    booking_id = Column(String, ForeignKey("bookings.id"), nullable=False)
    
    # Payment information
    amount = Column(Float, nullable=False)
    payment_method = Column(String, nullable=False)  # card, cash, bank_transfer, etc.
    payment_reference = Column(String, nullable=True)
    payment_gateway = Column(String, nullable=True)  # stripe, paypal, etc.
    
    # Status
    status = Column(String, default="pending", nullable=False)  # pending, completed, failed, refunded
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    processed_at = Column(DateTime, nullable=True)
    
    # Additional information
    notes = Column(Text, nullable=True)
    gateway_response = Column(Text, nullable=True)  # JSON string
    
    # Relationships
    booking = relationship("Booking")
    
    def __repr__(self):
        return f"<BookingPayment(id='{self.id}', amount={self.amount}, status='{self.status}')>"

class BookingModification(Base):
    """Booking modification history"""
    
    __tablename__ = "booking_modifications"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Booking relationship
    booking_id = Column(String, ForeignKey("bookings.id"), nullable=False)
    
    # Modification details
    modification_type = Column(String, nullable=False)  # date_change, room_change, guest_change, etc.
    old_values = Column(Text, nullable=True)  # JSON string
    new_values = Column(Text, nullable=True)  # JSON string
    reason = Column(Text, nullable=True)
    
    # ADDED: Missing fields from Pydantic
    new_check_out = Column(DateTime, nullable=True)
    new_check_in = Column(DateTime, nullable=True)
    new_room_id = Column(String, nullable=True)
    
    # Charges
    additional_charges = Column(Float, default=0.0, nullable=False)
    refund_amount = Column(Float, default=0.0, nullable=False)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    modified_by = Column(String, nullable=True)  # user_id who made the modification
    
    # Relationships
    booking = relationship("Booking")
    
    def __repr__(self):
        return f"<BookingModification(id='{self.id}', type='{self.modification_type}', booking_id='{self.booking_id}')>"

class BookingCancellation(Base):
    """Booking cancellation details"""
    
    __tablename__ = "booking_cancellations"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Booking relationship
    booking_id = Column(String, ForeignKey("bookings.id"), nullable=False)
    
    # Cancellation details
    reason = Column(Text, nullable=True)
    cancelled_by = Column(String, nullable=True)  # guest, admin, system
    cancellation_fee = Column(Float, default=0.0, nullable=False)
    refund_amount = Column(Float, default=0.0, nullable=False)
    
    # Timestamps
    cancelled_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    
    # Relationships
    booking = relationship("Booking")
    
    def __repr__(self):
        return f"<BookingCancellation(id='{self.id}', booking_id='{self.booking_id}', reason='{self.reason}')>"

class BookingGuest(Base):
    """Additional guests for booking"""
    
    __tablename__ = "booking_guests"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Booking relationship
    booking_id = Column(String, ForeignKey("bookings.id"), nullable=False)
    
    # Guest information
    guest_name = Column(String, nullable=False)
    guest_type = Column(String, nullable=False)  # adult, child, infant
    age = Column(Integer, nullable=True)
    special_requests = Column(Text, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    
    # Relationships
    booking = relationship("Booking")
    
    def __repr__(self):
        return f"<BookingGuest(id='{self.id}', name='{self.guest_name}', type='{self.guest_type}')>"

class BookingService(Base):
    """Additional services for booking"""
    
    __tablename__ = "booking_services"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Booking relationship
    booking_id = Column(String, ForeignKey("bookings.id"), nullable=False)
    
    # Service information
    service_name = Column(String, nullable=False)
    service_type = Column(String, nullable=False)  # spa, dining, transportation, etc.
    quantity = Column(Integer, default=1, nullable=False)
    unit_price = Column(Float, nullable=False)
    total_price = Column(Float, nullable=False)
    
    # Status
    status = Column(String, default="pending", nullable=False)  # pending, confirmed, completed, cancelled
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    scheduled_at = Column(DateTime, nullable=True)
    
    # Additional information
    notes = Column(Text, nullable=True)
    
    # Relationships
    booking = relationship("Booking")
    
    def __repr__(self):
        return f"<BookingService(id='{self.id}', service='{self.service_name}', booking_id='{self.booking_id}')>"

class BookingNote(Base):
    """Booking notes and comments"""
    
    __tablename__ = "booking_notes"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Booking relationship
    booking_id = Column(String, ForeignKey("bookings.id"), nullable=False)
    
    # Note information
    note_type = Column(String, nullable=False)  # general, guest_request, internal, system
    content = Column(Text, nullable=False)
    is_internal = Column(Boolean, default=False, nullable=False)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    created_by = Column(String, nullable=True)  # user_id who created the note
    
    # Relationships
    booking = relationship("Booking")
    
    def __repr__(self):
        return f"<BookingNote(id='{self.id}', type='{self.note_type}', booking_id='{self.booking_id}')>"

class BookingAudit(Base):
    """Booking audit trail"""
    
    __tablename__ = "booking_audits"
    
    # Primary key
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # Booking relationship
    booking_id = Column(String, ForeignKey("bookings.id"), nullable=False)
    
    # Audit information
    action = Column(String, nullable=False)  # created, updated, cancelled, checked_in, etc.
    old_values = Column(Text, nullable=True)  # JSON string
    new_values = Column(Text, nullable=True)  # JSON string
    
    # ADDED: Missing fields from Pydantic
    changed_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    changed_by = Column(String, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    created_by = Column(String, nullable=True)  # user_id who performed the action
    ip_address = Column(String, nullable=True)
    
    # Relationships
    booking = relationship("Booking")
    
    def __repr__(self):
        return f"<BookingAudit(id='{self.id}', action='{self.action}', booking_id='{self.booking_id}')>"

File: backend/src/entities/Booking.ts
/**
 * Booking Entity
 * TypeORM entity for booking management
 * Converted from Python SQLAlchemy model
 */

import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  OneToMany,
  JoinColumn,
  BeforeInsert,
} from 'typeorm';
import { v4 as uuidv4 } from 'uuid';
import { User } from './User';
import { Room } from './Room';
import { Payment } from './Payment';

@Entity('bookings')
export class Booking {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  // User relationship
  @Column({ name: 'user_id' })
  userId: string;

  // Room relationship
  @Column({ name: 'room_id' })
  roomId: string;

  // Property relationship
  @Column({ name: 'property_id' })
  propertyId: string;

  // Booking information
  @Column({ name: 'booking_reference' })
  bookingReference: string;

  @Column({ name: 'check_in_date', type: 'date' })
  checkInDate: Date;

  @Column({ name: 'check_out_date', type: 'date' })
  checkOutDate: Date;

  @Column({ name: 'check_in_time', type: 'time', nullable: true })
  checkInTime: string;

  @Column({ name: 'check_out_time', type: 'time', nullable: true })
  checkOutTime: string;

  // Guest information
  @Column({ name: 'guest_count' })
  guestCount: number;

  @Column({ name: 'guest_name', nullable: true })
  guestName: string;

  @Column({ name: 'guest_email', nullable: true })
  guestEmail: string;

  @Column({ name: 'guest_phone', nullable: true })
  guestPhone: string;

  @Column({ name: 'special_requests', type: 'text', nullable: true })
  specialRequests: string;

  // Pricing
  @Column({ name: 'base_price', type: 'decimal', precision: 10, scale: 2 })
  basePrice: number;

  @Column({ name: 'total_price', type: 'decimal', precision: 10, scale: 2 })
  totalPrice: number;

  @Column({ name: 'tax_amount', type: 'decimal', precision: 10, scale: 2, default: 0 })
  taxAmount: number;

  @Column({ name: 'discount_amount', type: 'decimal', precision: 10, scale: 2, default: 0 })
  discountAmount: number;

  @Column({ name: 'currency', default: 'NAD' })
  currency: string;

  // Booking status
  @Column({ name: 'status', default: 'pending' })
  status: string; // pending, confirmed, checked_in, checked_out, cancelled, no_show

  @Column({ name: 'payment_status', default: 'pending' })
  paymentStatus: string; // pending, paid, partially_paid, refunded, failed

  // Dates and times
  @Column({ name: 'confirmed_at', nullable: true })
  confirmedAt: Date;

  @Column({ name: 'checked_in_at', nullable: true })
  checkedInAt: Date;

  @Column({ name: 'checked_out_at', nullable: true })
  checkedOutAt: Date;

  @Column({ name: 'cancelled_at', nullable: true })
  cancelledAt: Date;

  @Column({ name: 'cancellation_reason', nullable: true })
  cancellationReason: string;

  // Timestamps
  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date;

  // Relationships
  @ManyToOne(() => User, (user) => user.bookings)
  @JoinColumn({ name: 'user_id' })
  user: User;

  @ManyToOne(() => Room, (room) => room.bookings)
  @JoinColumn({ name: 'room_id' })
  room: Room;

  @OneToMany(() => Payment, (payment) => payment.booking)
  payments: Payment[];

  @BeforeInsert()
  generateId() {
    if (!this.id) {
      this.id = uuidv4();
    }
    if (!this.bookingReference) {
      this.bookingReference = this.generateBookingReference();
    }
  }

  /**
   * Generate booking reference
   */
  private generateBookingReference(): string {
    const timestamp = Date.now().toString(36);
    const random = Math.random().toString(36).substring(2, 8);
    return `BK${timestamp}${random}`.toUpperCase();
  }

  /**
   * Convert booking to dictionary
   */
  toDict(): Record<string, any> {
    return {
      id: this.id,
      userId: this.userId,
      roomId: this.roomId,
      propertyId: this.propertyId,
      bookingReference: this.bookingReference,
      checkInDate: this.checkInDate?.toISOString().split('T')[0],
      checkOutDate: this.checkOutDate?.toISOString().split('T')[0],
      checkInTime: this.checkInTime,
      checkOutTime: this.checkOutTime,
      guestCount: this.guestCount,
      guestName: this.guestName,
      guestEmail: this.guestEmail,
      guestPhone: this.guestPhone,
      specialRequests: this.specialRequests,
      basePrice: this.basePrice,
      totalPrice: this.totalPrice,
      taxAmount: this.taxAmount,
      discountAmount: this.discountAmount,
      currency: this.currency,
      status: this.status,
      paymentStatus: this.paymentStatus,
      confirmedAt: this.confirmedAt?.toISOString(),
      checkedInAt: this.checkedInAt?.toISOString(),
      checkedOutAt: this.checkedOutAt?.toISOString(),
      cancelledAt: this.cancelledAt?.toISOString(),
      cancellationReason: this.cancellationReason,
      createdAt: this.createdAt?.toISOString(),
      updatedAt: this.updatedAt?.toISOString(),
    };
  }

  /**
   * Get booking display name
   */
  getDisplayName(): string {
    return `Booking ${this.bookingReference}`;
  }

  /**
   * Check if booking is pending
   */
  isPending(): boolean {
    return this.status === 'pending';
  }

  /**
   * Check if booking is confirmed
   */
  isConfirmed(): boolean {
    return this.status === 'confirmed';
  }

  /**
   * Check if booking is checked in
   */
  isCheckedIn(): boolean {
    return this.status === 'checked_in';
  }

  /**
   * Check if booking is checked out
   */
  isCheckedOut(): boolean {
    return this.status === 'checked_out';
  }

  /**
   * Check if booking is cancelled
   */
  isCancelled(): boolean {
    return this.status === 'cancelled';
  }

  /**
   * Check if booking is no show
   */
  isNoShow(): boolean {
    return this.status === 'no_show';
  }

  /**
   * Confirm booking
   */
  confirm(): void {
    this.status = 'confirmed';
    this.confirmedAt = new Date();
  }

  /**
   * Check in guest
   */
  checkIn(): void {
    this.status = 'checked_in';
    this.checkedInAt = new Date();
  }

  /**
   * Check out guest
   */
  checkOut(): void {
    this.status = 'checked_out';
    this.checkedOutAt = new Date();
  }

  /**
   * Cancel booking
   */
  cancel(reason?: string): void {
    this.status = 'cancelled';
    this.cancelledAt = new Date();
    if (reason) {
      this.cancellationReason = reason;
    }
  }

  /**
   * Mark as no show
   */
  markNoShow(): void {
    this.status = 'no_show';
  }

  /**
   * Check if payment is pending
   */
  isPaymentPending(): boolean {
    return this.paymentStatus === 'pending';
  }

  /**
   * Check if payment is paid
   */
  isPaymentPaid(): boolean {
    return this.paymentStatus === 'paid';
  }

  /**
   * Check if payment is partially paid
   */
  isPaymentPartiallyPaid(): boolean {
    return this.paymentStatus === 'partially_paid';
  }

  /**
   * Check if payment is refunded
   */
  isPaymentRefunded(): boolean {
    return this.paymentStatus === 'refunded';
  }

  /**
   * Check if payment failed
   */
  isPaymentFailed(): boolean {
    return this.paymentStatus === 'failed';
  }

  /**
   * Update payment status
   */
  updatePaymentStatus(status: string): void {
    this.paymentStatus = status;
  }

  /**
   * Calculate number of nights
   */
  getNumberOfNights(): number {
    const checkIn = new Date(this.checkInDate);
    const checkOut = new Date(this.checkOutDate);
    const timeDiff = checkOut.getTime() - checkIn.getTime();
    return Math.ceil(timeDiff / (1000 * 3600 * 24));
  }

  /**
   * Calculate total amount
   */
  getTotalAmount(): number {
    return this.totalPrice + this.taxAmount - this.discountAmount;
  }

  /**
   * Check if booking is active
   */
  isActive(): boolean {
    return ['confirmed', 'checked_in'].includes(this.status);
  }

  /**
   * Check if booking can be cancelled
   */
  canBeCancelled(): boolean {
    return ['pending', 'confirmed'].includes(this.status);
  }

  /**
   * Check if booking can be modified
   */
  canBeModified(): boolean {
    return ['pending', 'confirmed'].includes(this.status);
  }

  /**
   * Get booking summary
   */
  getSummary(): Record<string, any> {
    return {
      id: this.id,
      bookingReference: this.bookingReference,
      checkInDate: this.checkInDate?.toISOString().split('T')[0],
      checkOutDate: this.checkOutDate?.toISOString().split('T')[0],
      guestCount: this.guestCount,
      totalPrice: this.totalPrice,
      status: this.status,
      paymentStatus: this.paymentStatus,
      numberOfNights: this.getNumberOfNights(),
    };
  }

  /**
   * Check if booking is for today
   */
  isToday(): boolean {
    const today = new Date().toISOString().split('T')[0];
    return this.checkInDate?.toISOString().split('T')[0] === today;
  }

  /**
   * Check if booking is for tomorrow
   */
  isTomorrow(): boolean {
    const tomorrow = new Date();
    tomorrow.setDate(tomorrow.getDate() + 1);
    const tomorrowStr = tomorrow.toISOString().split('T')[0];
    return this.checkInDate?.toISOString().split('T')[0] === tomorrowStr;
  }

  /**
   * Check if booking is overdue (past check-out date)
   */
  isOverdue(): boolean {
    if (!this.checkOutDate) {
      return false;
    }
    return new Date() > this.checkOutDate && this.status !== 'checked_out';
  }
}